File: problems/array/0121-best-time-to-buy-and-sell-stock.py
"""#121 Best Time to Buy and Sell Stock
Track the lowest earlier price and the best profit seen so far.
Time: O(n) | Space: O(1)
"""


class Solution:
    def maxProfit(self, prices):
        lowest_price = float("inf")
        best_profit = 0

        for price in prices:
            lowest_price = min(lowest_price, price)
            best_profit = max(best_profit, price - lowest_price)

        return best_profit

File: problems/array/0121-best-time-to-buy-and-sell-stock.js
/**
 * #121 Best Time to Buy and Sell Stock
 * Pattern: one pass; track the lowest earlier price.
 * Invariant: minPrice is the lowest price before the current day.
 * Time: O(n) | Space: O(1)
 *
 * @param {number[]} prices
 * @return {number}
 */
function maxProfit(prices) {
  let minPrice = Infinity
  let bestProfit = 0

  for (const price of prices) {
    minPrice = Math.min(minPrice, price)
    bestProfit = Math.max(bestProfit, price - minPrice)
  }

  return bestProfit
}

module.exports = maxProfit
